### ISSUE TITLE
Support Template Haskell Typed Quotes

### ISSUE THREAD

**Issue by MHova** [ndmitchell/hlint#332]
This is probably a `haskell-src-exts` issue, but opening here anyway.

HLint is missing support for [typed quotes](https://ghc.haskell.org/trac/ghc/wiki/TemplateHaskell/Typed), which has the `[||...||]` syntax. Currently HLint complains with a parse error.

**Comment by ndmitchell** [comment]
Thanks for the info - almost certainly one to report to haskell-src-exts. Do you have a small file using typed quotes that works as a valid GHC program (e.g. `runhaskell` on it works) but fails to parse? That's what the HSE people tend to use for reproducing bugs.

**Comment by ndmitchell** [comment]
Did this get into a real GHC release? I can't find it in the user manual.

**Comment by MHova** [comment]
Are you talking about a fix in `haskell-src-exts` to support Typed Quotes? Or the existence of the Typed Quotes feature at all in GHC?

**Comment by ndmitchell** [comment]
The feature in GHC. I would usually go to the GHC manual, grab a small sample, and then create an HSE ticket containing that. However, in this instance, googling just took me to the wiki.

**Comment by MHova** [comment]
Sorry for the ridiculously delayed response on this, but here's a simple example showing the problem.

```haskell
{-# LANGUAGE TemplateHaskell #-}

module Example where

import Language.Haskell.TH (TExp, Q, runIO)

example :: Q (TExp String)
example = [|| "whatever" ||]
```

This code snippet compiles, but `hlint` can't parse the code because:

```
Example.hs:8:13: Error: Parse error
Found:
    example :: Q (TExp String)
  > example = [|| "whatever" ||]
```

Using `GHC 7.10.3` and `hlint 2.0.9`.

**Comment by ndmitchell** [comment]
Thanks, reported upstream at https://github.com/haskell-suite/haskell-src-exts/issues/383.

**Comment by ndmitchell** [comment]
Confirmed works with HEAD that doesn't use HSE for parsing, so fixed in the next release.
